gui

```diff
@@ -1,28 +1,280 @@
+      #if disabled, display only an 'enable' button
+      whichbutton+=("--button=<b>Enable</b>!!Force this app to install on your system."$'\n'"This app was disabled for a reason so if you enable it..."$'\n'"YOU HAVE BEEN WARNED.:12")
+    else
+      #if app status is 'corrupted', and a log file exists for this app, then display a button to view the log file
+      if [ "$app_status" == 'corrupted' ];then
+        logfile="$(ls "$DIRECTORY/logs"/* -t | grep "fail-${app}" | head -n1)"
+        if [ ! -z "$logfile" ];then
+          whichbutton+=("--button=Errors!${DIRECTORY}/icons/log-file.png!$app failed to $(echo "$(basename "$logfile")" | awk -F'-' '{print $1}'). Click this button to view the error output saved in the log file.:14")
+        fi
+      fi
+      #if status is corrupted or unknown, then show both buttons
+      whichbutton+=("--button=Uninstall!${DIRECTORY}/icons/uninstall.png:2" "--button=Install!${DIRECTORY}/icons/install.png:4")
+    fi
+    
+    #only display app icon if it exists
+    if [ -f "${DIRECTORY}/apps/${app}/icon-64.png" ];then
+      imageline=(--image="${DIRECTORY}/apps/${app}/icon-64.png" --image-on-top)
+    fi
+    
+    (cat "${DIRECTORY}/apps/$app/description" || echo "Description unavailable") | yad --text-info --fontname=12 --wrap --show-uri \
+      --text="$(sed 's/&/&amp;/g' <<<"$abovetext")" \
+      "${imageline[@]}" \
+      --title="Details of ${app}" --window-icon="${DIRECTORY}/icons/logo.png" \
+      $geometry2 --close-on-unfocus --skip-taskbar \
+      "${whichbutton[@]}"
+    button=$?
+    
+    case $button in
+    4)
+      #install
+      terminal_manage install "$app"
+      ;;
+    2)
+      #uninstall
+      terminal_manage uninstall "$app"
+      ;;
+    6)
+      #scripts
+      install_script="${DIRECTORY}/apps/${app}/$(script_name_cpu "$app")"
+      uninstall_script="${DIRECTORY}/apps/${app}/uninstall"
+      
+      text_editor "$uninstall_script" &
+      sleep 0.1
+      text_editor "$install_script" &
+      
+      "${DIRECTORY}/gui" "$@" #open details window again
+      ;;
+    8)
+      #edit
+      "${DIRECTORY}/createapp" "$app"
+      refresh_list
+      ;;
+    12)
+      #enable
+      
+      #remove status file containing 'disabled'
+      rm -f "${DIRECTORY}/data/status/${app}"
+      refresh_list &
+      "${DIRECTORY}/gui" "$@" #open details window again
+      ;;
+    14)
+      #viewlog
+      echo "Viewing error log of $app..."
+      echo "Log filepath: $logfile"
+      "${DIRECTORY}/etc/viewlog" "$logfile"
+      
+      "${DIRECTORY}/gui" "$@" #open details window again
+      ;;
+    *)
+      #unknown button
+      exit 0
+      ;;
+    esac
+    
+  fi
+  exit 0
 fi
 
-#in background, check mission-critical scripts and re-download them if they contain a syntax error
+DIRECTORY="$(readlink -f "$(dirname "$0")")"
+
+set -a #make all functions in the api available to subprocesses
+source "${DIRECTORY}/api" || error "failed to source ${DIRECTORY}/api"
```

Comment:
same issues as before, if the api fails to source, then the error function isn't available so it won't execute
```suggestion
source "${DIRECTORY}/api" || (  echo -e "\e[91mfailed to source ${DIRECTORY}/api\e[0m" 1>&2; exit 1 )
```

Reply:
I think that putting `exit 1` in parenthesis won't cause the script to exit, so I'll just add back the `error()` function to the gui script.

Reply:
ah right yes it won't, putting the error function at the top is the easiest solution, you can do that